tasks/static_quality_gates/gates.py

```diff
@@ -0,0 +1,866 @@
+        if not self.arch:
+            raise ValueError("arch cannot be empty")
+        if not self.os:
+            raise ValueError("os cannot be empty")
+
+
+@dataclass(frozen=True)
+class SizeViolation:
+    """Represents a size limit violation"""
+
+    measurement_type: str  # "wire" or "disk"
+    current_size: int
+    max_size: int
+
+    @property
+    def excess_bytes(self) -> int:
+        """Number of bytes over the limit"""
+        return self.current_size - self.max_size
+
+    @property
+    def excess_mb(self) -> float:
+        """Number of MB over the limit"""
+        return self.excess_bytes / (1024 * 1024)
+
+
+@dataclass(frozen=True)
+class GateResult:
+    """
+    Result of executing a quality gate.
+    Contains all information needed for reporting.
+    """
+
+    config: QualityGateConfig
+    measurement: ArtifactMeasurement
+    violations: list[SizeViolation]
+    success: bool
+
+    @property
+    def wire_remaining_bytes(self) -> int:
+        """Remaining wire size capacity in bytes"""
+        return max(0, self.config.max_on_wire_size - self.measurement.on_wire_size)
+
+    @property
+    def disk_remaining_bytes(self) -> int:
+        """Remaining disk size capacity in bytes"""
+        return max(0, self.config.max_on_disk_size - self.measurement.on_disk_size)
+
+
+class ArtifactMeasurer(Protocol):
+    """
+    Protocol for measuring artifacts.
+    Implementations handle specific artifact types (Docker, packages, etc.)
+    """
+
+    def measure(self, ctx: Context, config: QualityGateConfig) -> ArtifactMeasurement:
+        """
+        Measure an artifact's on-wire and on-disk sizes.
+
+        Args:
+            ctx: Invoke context for running commands
+            config: Quality gate configuration
+
+        Returns:
+            ArtifactMeasurement with populated sizes
+
+        Raises:
+            StaticQualityGateFailed: If measurement fails
+            InfraError: If there's an infrastructure issue (retryable)
+        """
+        ...
+
+
+class PackageArtifactMeasurer:
+    """
+    Measures package artifacts (DEB, RPM, MSI, etc.).
+    """
+
+    def measure(self, ctx: Context, config: QualityGateConfig) -> ArtifactMeasurement:
+        """Measure package artifact sizes"""
+        try:
+            artifact_paths = self._find_package_paths(config)
+            wire_size, disk_size = self._calculate_package_sizes(ctx, config, artifact_paths)
+
+            # For packages, the primary artifact path is the main package file
+            primary_path = artifact_paths.get('primary', artifact_paths.get('msi', ''))
+
+            return ArtifactMeasurement(artifact_path=primary_path, on_wire_size=wire_size, on_disk_size=disk_size)
+        except Exception as e:
+            if isinstance(e, StaticQualityGateFailed | InfraError):
+                raise
+            raise StaticQualityGateFailed(f"Failed to measure package {config.gate_name}: {e}") from e
```

Comment:
The following adjustment would leverage Python’s exception hierarchy and control flow:
```suggestion
        except (StaticQualityGateFailed, InfraError):
            raise
        except Exception as e:
            raise StaticQualityGateFailed(f"Failed to measure package {config.gate_name}") from e
```
Also: no `: {e}` in the message: avoids redundancy, because `from e` already includes the underlying exception in the output.